iocore/net/QUICPacketHandler.cc

```diff
@@ -0,0 +1,148 @@
+  na  = new QUICPacketHandler(opt, this->_ssl_ctx);
+  *na = *this;
+  return na;
+}
+
+int
+QUICPacketHandler::acceptEvent(int event, void *data)
+{
+  // NetVConnection *netvc;
+  ink_release_assert(event == NET_EVENT_DATAGRAM_OPEN || event == NET_EVENT_DATAGRAM_READ_READY ||
+                     event == NET_EVENT_DATAGRAM_ERROR);
+  ink_release_assert((event == NET_EVENT_DATAGRAM_OPEN) ? (data != nullptr) : (1));
+  ink_release_assert((event == NET_EVENT_DATAGRAM_READ_READY) ? (data != nullptr) : (1));
+
+  if (event == NET_EVENT_DATAGRAM_OPEN) {
+    // Nothing to do.
+    return EVENT_CONT;
+  } else if (event == NET_EVENT_DATAGRAM_READ_READY) {
+    Queue<UDPPacket> *queue = (Queue<UDPPacket> *)data;
+    UDPPacket *packet_r;
+    ip_port_text_buffer ipb;
+    while ((packet_r = queue->dequeue())) {
+      Debug("quic_sec", "received packet from %s, size=%lld", ats_ip_nptop(&packet_r->from.sa, ipb, sizeof(ipb)),
+            packet_r->getPktLength());
+      this->_recv_packet(event, packet_r);
+    }
+    return EVENT_CONT;
+  }
+
+  /////////////////
+  // EVENT_ERROR //
+  /////////////////
+  if (((long)data) == -ECONNABORTED) {
+  }
+
+  ink_abort("QUIC accept received fatal error: errno = %d", -((int)(intptr_t)data));
+  return EVENT_CONT;
+  return 0;
+}
+
+void
+QUICPacketHandler::init_accept(EThread *t = nullptr)
+{
+  SET_HANDLER(&QUICPacketHandler::acceptEvent);
+}
+
+void
+QUICPacketHandler::_recv_packet(int event, UDPPacket *udpPacket)
+{
+  IOBufferBlock *block = udpPacket->getIOBlockChain();
+
+  std::unique_ptr<QUICPacket> qPkt = std::unique_ptr<QUICPacket>(QUICPacketFactory::create(block));
+  QUICNetVConnection *vc           = this->_connections.get(qPkt->connection_id());
+
+  if (!vc) {
+    // Unknown Connection ID
+    Connection con;
+    con.setRemote(&udpPacket->from.sa);
+    vc =
+      static_cast<QUICNetVConnection *>(getNetProcessor()->allocate_vc(((UnixUDPConnection *)udpPacket->getConnection())->ethread));
+    vc->init(udpPacket->getConnection(), this);
+    vc->id = net_next_connection_number();
+    vc->con.move(con);
+    vc->submit_time = Thread::get_hrtime();
+    vc->mutex       = this->mutex;
+    vc->action_     = *this->action_;
+    vc->set_is_transparent(this->opt.f_inbound_transparent);
+    vc->set_context(NET_VCONNECTION_IN);
+    vc->read.triggered = 1;
+    vc->start(this->_ssl_ctx);
+    // TODO: Handle Connection ID of Client Cleartext / Non-Final Server Cleartext Packet
+    this->_connections.put(qPkt->connection_id(), vc);
+  }
+
+  vc->push_packet(std::move(qPkt));
+
+  // send to EThread
+  eventProcessor.schedule_imm(vc, ET_CALL, QUIC_EVENT_PACKET_READ_READY, nullptr);
+}
+
+// TODO: Should be called via eventProcessor?
+void
+QUICPacketHandler::send_packet(const QUICPacket &packet, QUICNetVConnection *vc)
+{
+  // TODO: remove a connection which is created by Client Initial
+  //       or update key to new one
+  if (!this->_connections.get(packet.connection_id())) {
```

Comment:
Seems like there might be contention on connections.  We might want to think about how to reduce this later.

Reply:
This is using a Map to store the connection.  As I remember this it not thread safe and you will have to have a lock to access it.  It will be expensive to access it.

Reply:
I see. I created a issue for that. #2478